docs/shadow_config_dynamic.md

```diff
@@ -0,0 +1,307 @@
+1.2](https://yaml.org/spec/1.2.2/) format, there are many options
+available; even more so since [JSON](https://en.wikipedia.org/wiki/JSON) is
+also valid YAML 1.2.
+
+## YAML templating
+
+YAML itself has some features to help avoid repetition. When using these
+features, it can be helpful to use shadow's `--show-config` flag to examine the
+"flat" generated config.
+
+An individual node can be made into an *anchor* (`&AnchorName x`), and
+referenced via an *alias* (`*AnchorName`). For example, here we create
+and use the anchors `Node`, `Fast`, `Slow`, `FooPath`, and `BarPath`:
+
+```yaml
+general:
+  stop_time: 10s
+network:
+  graph:
+    type: 1_gbit_switch
+hosts:
+  fast_foo:
+    network_node_id: &Node 0
+    bandwidth_up: &Fast "100 Mbit"
+    bandwidth_down: *Fast
+    processes:
+    - path: &FooPath "/path/to/foo"
+    # ...
+  slow_foo:
+    network_node_id: *Node
+    bandwidth_up: &Slow "1 Mbit"
+    bandwidth_down: *Slow
+    processes:
+    - path: *FooPath
+    # ...
+  fast_bar:
+    network_node_id: *Node
+    bandwidth_up: *Fast
+    bandwidth_down: *Fast
+    processes:
+    - path: &BarPath "/path/to/bar"
+    # ...
+  slow_bar:
+    network_node_id: *Node
+    bandwidth_up: *Slow
+    bandwidth_down: *Slow
+    processes:
+    - path: *BarPath
+```
+
+We can also enable shadow's experimental
+[extended yaml](./shadow_config_spec.md#experimentaluse_extended_yaml)
+support, allowing us to use [merge keys](https://yaml.org/type/merge.html)
+and [extension
+fields](https://docs.docker.com/compose/compose-file/#extension)).
+
+We can use extension fields to move our constants into one place:
+
+```yaml
+x-constants:
+  - &Node 0
+  - &Fast "100 Mbit"
+  - &Slow "1 Mbit"
+  - &FooPath "/path/to/foo"
+  - &BarPath "/path/to/bar"
+general:
+  stop_time: 10s
+network:
+  graph:
+    type: 1_gbit_switch
+hosts:
+  fast_foo:
+    network_node_id: *Node
+    bandwidth_up: *Fast
+    bandwidth_down: *Fast
+    processes:
+    - path: *FooPath
+  slow_foo:
+    network_node_id: *Node
+    bandwidth_up: *Slow
+    bandwidth_down: *Slow
+    processes:
+    - path: *FooPath
+  fast_bar:
+    network_node_id: *Node
+    bandwidth_up: *Fast
+    bandwidth_down: *Fast
+    processes:
+    - path: *BarPath
+  slow_bar:
+    network_node_id: *Node
+    bandwidth_up: *Slow
+    bandwidth_down: *Slow
+    processes:
+    - path: *BarPath
+```
+
+We can combine this with merge-keys to make extendable templates for fast and slow hosts:
+
+```yaml
+x-constants:
+  - &Node 0
+  - &Fast "100 Mbit"
+  - &Slow "1 Mbit"
+  - &FooPath "/path/to/foo"
+  - &BarPath "/path/to/bar"
+  - &FastHost
+    network_node_id: *Node
+    bandwidth_up: *Fast
+    bandwidth_down: *Fast
+  - &SlowHost
+    network_node_id: *Node
+    bandwidth_up: *Slow
+    bandwidth_down: *Slow
+general:
+  stop_time: 10s
+network:
+  graph:
+    type: 1_gbit_switch
+hosts:
+  fast_foo:
+    <<: *FastHost
+    processes:
+    - path: *FooPath
```

Comment:
One suggestion is that I find examples with a lot of indirection (for example with stuff like anchors and merging) to be easier to understand when they use concrete things instead of abstract things. So instead of "foo", to be a server or something. When using abstract names like "foo", it can be hard to relate it to a situation where these features would be useful in the real world, and it takes some mental energy to remember what a "foo" is (for example in this case "foo" is both a host name and a process path).

Reply:
I'd originally started with something more concrete using curl clients and python web servers, but that brought in too much incidental complexity.

I search-replaced foo -> client and bar -> server to make it a little more concrete, without making it too weird that I'm glossing over e.g. host names and ports.